Bio/SeqIO/XdnaIO.py

```diff
@@ -0,0 +1,296 @@
+    qualifiers = {}
+    if name:
+        qualifiers['label'] = [name]
+    _parse_feature_description(desc, qualifiers)
+    feature = SeqFeature(location, type=type, qualifiers=qualifiers)
+    record.features.append(feature)
+
+
+def XdnaIterator(handle):
+    """Parse a Xdna file and return a SeqRecord object.
+
+    Note that this is an "iterator" in name only since a Xdna file always
+    contain a single sequence.
+    """
+    # Parse fixed-size header and do some rudimentary checks
+    #
+    # The "neg_length" value is the length of the part of the sequence
+    # before the nucleotide considered as the "origin" (nucleotide number 1,
+    # which in DNA Strider is not always the first nucleotide).
+    # Biopython's SeqRecord has no such concept of a sequence origin as far
+    # as I know, so we ignore that value. SerialCloner has no such concept
+    # either and always generates files with a neg_length of zero.
+    header = _read(handle, 112)
+    (version, type, topology, length, neg_length, com_length) = unpack('>BBB25xII60xI12x', header)
+    if version != 0:
+        raise ValueError("Unsupported XDNA version")
+    if type not in _seq_types:
+        raise ValueError("Unknown sequence type")
+
+    # Read actual sequence and comment found in all XDNA files
+    sequence = _read(handle, length).decode('ASCII')
+    comment = _read(handle, com_length).decode('ASCII')
+
+    # Try to derive a name from the first "word" of the comment
+    name = comment.split(' ')[0]
+
+    # Create record object
+    record = SeqRecord(Seq(sequence, _seq_types[type]),
+                       description=comment, name=name, id=name)
+    if topology in _seq_topologies:
+        record.annotations['topology'] = _seq_topologies[topology]
+
+    if len(handle.read(1)) == 1:
+        # This is an XDNA file with an optional annotation section.
+
+        # Skip the overhangs as I don't know how to represent
+        # them in the SeqRecord model.
+        _read_overhang(handle)  # right-side overhang
+        _read_overhang(handle)  # left-side overhang
+
+        # Read the features
+        num_features = unpack('>B', _read(handle, 1))[0]
+        while num_features > 0:
+            _read_feature(handle, record)
+            num_features -= 1
+
+    yield record
+
+
+class XdnaWriter(SequenceWriter):
+    """Write files in the Xdna format."""
+
+    def write_file(self, records):
+        """Write the specified record to a Xdna file.
+
+        Note that the function expects a list of records as per the
+        SequenceWriter interface, but the list should contain only one
+        record as the Xdna format is a mono-record format.
+        """
+        if not records:
+            raise ValueError("Must have one sequence")
+        if len(records) > 1:
+            raise ValueError("More than one sequence found")
+
+        record = records[0]
+
+        alptype = type(record.seq.alphabet)
```

Comment:
This ought to work, but the alphabet system is too complicated for its own good. See this example for how it has to be done:

https://github.com/biopython/biopython/blob/biopython-174/Bio/SeqIO/SeqXmlIO.py#L348

This is one of the reasons we're looking at a massive simplification - see #2046.